### PR TITLE
refactor: port to macos

### PR DESCRIPTION
**Pull request by killagu**

### CODE REVIEW

**Comment by killagu** [comment]
current progress:
- run with macfuse succeed
- micro_http not work well, can not get requests

**Comment by killagu** [comment]
## Remove the event-manger dep
event-manager is linux specific, it use to trigger fuse session, http_thread exit.

In this refactor impl, wait fuse channel thread exit then exit the http_thread.

## Remove the eventfd dep
api-server use eventfd as the request notification to trigger server run. Because of remove event-manger and eventfd, it can work work.

Add new thread named api-server for it do event loop by self, and refactor api channel to `Option`. If api-server receiver none request means for exit.

File: storage/src/remote/sock_ctrl_msg.rs [hunk 1]
```diff
@@ -0,0 +1,753 @@
+use vmm_sys_util::errno::{Error as SysError, Result as SysResult};
```
Comment (killagu) [inline storage/src/remote/sock_ctrl_msg.rs:1]: Copy from https://github.com/xujihui1985/db-micro-http/blob/feat/cross-platform-support/src/common/sock_ctrl_msg.rs. It should pub new crate.

File: Cargo.toml [hunk 2]
```diff
@@ -47,8 +47,7 @@ openssl-src = ">=111.16.0"
 # pin rand_core to bring in fix for RUSTSEC-2021-0023
 rand_core = ">=0.6.2"
 
-event-manager = "0.2.1"
-fuse-backend-rs = { version = "0.3.0", optional = true }
+fuse-backend-rs = { git = "https://github.com/cloud-hypervisor/fuse-backend-rs", rev = "765d987642dfea1a5c25691a506b657827320cad", optional = true }
```
[the remark this answers is not in the document]
Comment (bergwolf) [reply Cargo.toml:50]: yup, I'll look into it.

File: Makefile [hunk 3]
```diff
@@ -86,6 +86,11 @@ ifdef NYDUS_TEST_VIRTIOFS
 	RUST_BACKTRACE=1 cargo test $(VIRIOFS_COMMON) --bin nydusd -- --nocapture --test-threads=8
 endif
 
+macos-ut:
```
Comment (bergwolf) [inline Makefile:89]: Can you enable the smoke test on macos as well? Just so that we know the code actually works roughly.

File: api/Cargo.toml [hunk 4]
```diff
@@ -10,11 +10,12 @@ license = "Apache-2.0 OR BSD-3-Clause"
 [dependencies]
 lazy_static = "1.4.0"
 log = "0.4.8"
-micro_http = { git = "https://github.com/cloud-hypervisor/micro-http.git", branch = "master" }
+micro_http = { git = "https://github.com/openanolis/dbs-uhttp", rev = "52e7ffe994134c231ad9fe1cc88c68b0cdda2278" }
```
Comment (bergwolf) [inline api/Cargo.toml:13]: @xujihui1985 @jiangliu could you release `dbs-uhttp` on cargo.io?

File: rafs/src/fs.rs [hunk 5]
```diff
@@ -14,6 +14,10 @@
 //! data. There are also [FsPrefetchControl](struct.FsPrefetchControl.html) and
 //! [RafsConfig](struct.RafsConfig.html) to configure an [Rafs] instance.
 
+#[cfg(target_os = "macos")]
+use libc::{stat as stat64, statvfs as statvfs64};
+#[cfg(target_os = "linux")]
+use libc::{stat64, statvfs64};
```
Comment (bergwolf) [inline rafs/src/fs.rs:20]: please use the wraper exports in `fuse-backend-rs::fuse_abi`

File: storage/src/backend/localfs.rs [hunk 6]
```diff
@@ -435,6 +437,37 @@ impl Prefetcher {
...
+                .checked_add(len as u64)
+                .ok_or_else(|| einval!("invalid length"))?;
+            if offset > self.blob_size || end > self.blob_size || zero != 0 {
+                return Err(einval!(format!(
+                    "invalid readahead entry ({}, {}), blob size {}",
+                    offset, len, self.blob_size
+                )));
+            }
+
+            unsafe {
```
Comment (bergwolf) [inline storage/src/backend/localfs.rs:455]: How about using `posix_fadvise` that is supposed to work on both Linux and Unix (macos).
https://man7.org/linux/man-pages/man2/posix_fadvise.2.html

File: storage/src/remote/connection.rs [hunk 7]
```diff
@@ -446,12 +453,14 @@ impl Endpoint {
...
     pub fn recv_data(&mut self, len: usize) -> Result<(usize, Vec<u8>)> {
         let mut rbuf = vec![0u8; len];
-        let mut iovs = [iovec {
-            iov_base: rbuf.as_mut_ptr() as *mut c_void,
-            iov_len: len,
-        }];
-        // Safe because we own rbuf and it's safe to fill a byte array with arbitrary data.
-        let (bytes, _) = unsafe { self.sock.recv_with_fds(&mut iovs, &mut [])? };
+        let (bytes, _) = {
+            let mut iovs = [iovec {
```
Comment (bergwolf) [inline storage/src/remote/connection.rs:457]: why is it moved? To drop early?

File: storage/src/utils.rs [hunk 8]
```diff
@@ -197,6 +201,26 @@ pub fn readahead(fd: libc::c_int, mut offset: u64, end: u64) {
     }
 }
 
+#[cfg(target_os = "macos")]
+pub fn readahead(fd: libc::c_int, mut offset: u64, end: u64) {
```
Comment (bergwolf) [inline storage/src/utils.rs:205]: Same here, could you unify them with posix_fadvise?

**Review by bergwolf — COMMENTED** [review]
Generally looks good. A few comments inline.  Thanks!

File: api/Cargo.toml (hunk 4 above)
Comment (jiangliu) [reply api/Cargo.toml:13]: We are waiting for a confirmation signal to publish "dbs-uhttp". Is it ready now?

File: Makefile (hunk 3 above)
Comment (killagu) [reply Makefile:89]: As the fuse-backend-rs, github action with macos can not use macfuse, so the smoke test can not run.

Comment (bergwolf) [reply Makefile:89]: Ok, fair enough. Let's just keep it w/o e2e tests and hopefully the UT can cover most cross-platform differences.

File: Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (bergwolf) [reply Cargo.toml:52]: @killagu could you verify if the latest fuse-backend-rs master works for you? Once you are ok with it, we can tag a new fuse-backend-rs release.

[the remark this answers is not in the document]
Comment (killagu) [reply Cargo.toml:52]: fuse-backend-rs master is works well with macos. But on linux, the syscall will hang. I'm woking on it.

File: storage/src/backend/localfs.rs (hunk 6 above)
Comment (killagu) [reply storage/src/backend/localfs.rs:482]: The `posix_fadvise`  not found in macos. It's sad..

File: storage/src/remote/connection.rs (hunk 7 above)
Comment (killagu) [reply storage/src/remote/connection.rs:457]: Unnecessary update. I'll revert it.

File: api/Cargo.toml (hunk 4 above)
Comment (killagu) [reply api/Cargo.toml:13]: @jiangliu It's ready, test for macos is ok.

File: Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (killagu) [reply Cargo.toml]: @bergwolf It's ready, test for macos is ok.

**Comment by bergwolf** [comment]
@killagu Could you rebase to the latest master to pick up some latest fix to the Makefile so that `cargo clippy` warnings are actually caught? Right now I see a lot of them showing up in CI, e.g.,
https://github.com/dragonflyoss/image-service/runs/5716304211?check_suite_focus=true

**Comment by bergwolf** [comment]
LGTM! Thanks a lot @killagu ! 

@xujihui1985 @jiangliu Please publish a `micro_http` version. I'll publish the `fuse-backend-rs` one. Then @killagu can update a bit to use the crates.io packages and this is ready to be merged. Thank you all!

**Comment by jiangliu** [comment]
> LGTM! Thanks a lot @killagu !
> 
> @xujihui1985 @jiangliu Please publish a `micro_http` version. I'll publish the `fuse-backend-rs` one. Then @killagu can update a bit to use the crates.io packages and this is ready to be merged. Thank you all!

published dbs-uhttp v0.2.0

**Comment by jiangliu** [comment]
> LGTM! Thanks a lot @killagu !
> 
> @xujihui1985 @jiangliu Please publish a `micro_http` version. I'll publish the `fuse-backend-rs` one. Then @killagu can update a bit to use the crates.io packages and this is ready to be merged. Thank you all!

published dbs-uhttp v0.2.0

**Comment by jiangliu** [comment]
> LGTM! Thanks a lot @killagu !
> 
> @xujihui1985 @jiangliu Please publish a `micro_http` version. I'll publish the `fuse-backend-rs` one. Then @killagu can update a bit to use the crates.io packages and this is ready to be merged. Thank you all!

published dbs-uhttp v0.2.0

**Comment by jiangliu** [comment]
> LGTM! Thanks a lot @killagu !
> 
> @xujihui1985 @jiangliu Please publish a `micro_http` version. I'll publish the `fuse-backend-rs` one. Then @killagu can update a bit to use the crates.io packages and this is ready to be merged. Thank you all!

published dbs-uhttp v0.2.0

**Comment by jiangliu** [comment]
> LGTM! Thanks a lot @killagu !
> 
> @xujihui1985 @jiangliu Please publish a `micro_http` version. I'll publish the `fuse-backend-rs` one. Then @killagu can update a bit to use the crates.io packages and this is ready to be merged. Thank you all!

published dbs-uhttp v0.2.0

**Comment by jiangliu** [comment]
> LGTM! Thanks a lot @killagu !
> 
> @xujihui1985 @jiangliu Please publish a `micro_http` version. I'll publish the `fuse-backend-rs` one. Then @killagu can update a bit to use the crates.io packages and this is ready to be merged. Thank you all!

published dbs-uhttp v0.2.0

**Comment by bergwolf** [comment]
`fuse-backend-rs` v0.4.0 has been published too. @killagu

**Comment by killagu** [comment]
@bergwolf `fuse-backend-rs` and `dbs-uhttp` all used from crates.

**Review by bergwolf — APPROVED** [review]
LGTM!

**Comment by bergwolf** [comment]
@killagu there are some clippy errors. Please take a look. Then this can be merged. Thanks a lot!

**Review by jiangliu — APPROVED** [review]

**Comment by killagu** [comment]
@bergwolf All checks have passed. 🎉

**Pull request merged**